src/ecdsapod.rs

```diff
@@ -0,0 +1,546 @@
+    builder: &mut CircuitBuilder<F, D>,
+    params: &Params,
+    msg: &[Target],
+    pk: &[Target],
+) -> Vec<StatementTarget> {
+    let st_type = StatementTarget::from_flattened(
+        params,
+        &builder.constants(&type_statement().to_fields(params)),
+    );
+
+    let ak_podid = builder.constant_value(RawValue::from(SELF.0));
+
+    let msg_hash = builder.hash_n_to_hash_no_pad::<PoseidonHash>(msg.to_vec());
+    let ak_key = builder.constant_value(Key::from(KEY_SIGNED_MSG).raw());
+    let ak_msg = StatementArgTarget::anchored_key(builder, &ak_podid, &ak_key);
+    let value = StatementArgTarget::literal(builder, &ValueTarget::from_slice(&msg_hash.elements));
+    let st_msg =
+        StatementTarget::new_native(builder, params, NativePredicate::ValueOf, &[ak_msg, value]);
+
+    let pk_hash = builder.hash_n_to_hash_no_pad::<PoseidonHash>(pk.to_vec());
+    let ak_key = builder.constant_value(Key::from(KEY_ECDSA_PK).raw());
+    let ak_pk = StatementArgTarget::anchored_key(builder, &ak_podid, &ak_key);
+    let value = StatementArgTarget::literal(builder, &ValueTarget::from_slice(&pk_hash.elements));
+    let st_pk =
+        StatementTarget::new_native(builder, params, NativePredicate::ValueOf, &[ak_pk, value]);
+
+    vec![st_type, st_msg, st_pk]
+}
+
+// compatible with the same method in-circuit (pub_self_statements_target)
+fn pub_self_statements(
+    msg: Secp256K1Scalar,
+    pk: ECDSAPublicKey<Secp256K1>,
+) -> Vec<middleware::Statement> {
+    let st_type = type_statement();
+
+    // hash the msg as in-circuit
+    let msg_limbs = scalar_to_limbs(msg);
+    let msg_hash = PoseidonHash::hash_no_pad(&msg_limbs);
+
+    let st_msg = Statement::ValueOf(
+        AnchoredKey::from((SELF, KEY_SIGNED_MSG)),
+        Value::from(RawValue(msg_hash.elements)),
+    );
+
+    // hash the pk as in-circuit
+    let pk_x_limbs = base_to_limbs(pk.0.x);
+    let pk_y_limbs = base_to_limbs(pk.0.y);
+    let pk_hash = PoseidonHash::hash_no_pad(&[pk_x_limbs, pk_y_limbs].concat());
+
+    let st_pk = Statement::ValueOf(
+        AnchoredKey::from((SELF, KEY_ECDSA_PK)),
+        Value::from(RawValue(pk_hash.elements)),
+    );
+
+    vec![st_type, st_msg, st_pk]
+}
+
+fn scalar_to_limbs(v: Secp256K1Scalar) -> Vec<F> {
+    let max_num_limbs = 8;
+    let v_biguint = biguint_from_array(std::array::from_fn(|i| v.0[i]));
+    let mut limbs = v_biguint.to_u32_digits();
+    assert!(max_num_limbs >= limbs.len());
+    limbs.resize(max_num_limbs, 0);
+    let limbs_f: Vec<F> = limbs.iter().map(|l| F::from_canonical_u32(*l)).collect();
+    limbs_f
+}
+
+fn base_to_limbs(v: Secp256K1Base) -> Vec<F> {
+    let max_num_limbs = 8;
+    let v_biguint = biguint_from_array(std::array::from_fn(|i| v.0[i]));
+    let mut limbs = v_biguint.to_u32_digits();
+    assert!(max_num_limbs >= limbs.len());
+    limbs.resize(max_num_limbs, 0);
+    let limbs_f: Vec<F> = limbs.iter().map(|l| F::from_canonical_u32(*l)).collect();
+    limbs_f
+}
```

Comment:
this is a minor thing but these two functions have the same code.  Both `Secp256K1Scalar` and `Secp256K1Base` are `[u64; 4]` so maybe you can deduplicate the code?